### ISSUE TITLE
Add binary agent for LicheeRV nano board

### ISSUE THREAD

**Issue by leleb** [Ylianst/MeshAgent#266]
I just got one of these remote KVM:
https://github.com/sipeed/NanoKVM
https://wiki.sipeed.com/hardware/en/kvm/NanoKVM/introduction.html
and it would be very usefull if meshcentral agent could be installed in it.
Tried al binaries but none working for this riscv device.
It is based on LicheeRV nano board ( https://wiki.sipeed.com/hardware/en/lichee/RV_Nano/1_intro.html )

**Comment by si458** [comment]
i have just moved the issue to `meshagent` as its to do that rather than `meshcentral` 👍 
once we have a binary built that works with it, we can then add the option into `meshcentral`

**Comment by si458** [comment]
That has to be the quickest I've ordered something off amazon UK, had it posted from China and arrive in the UK! 10 days in transit literally! 
![20241223_183140](https://github.com/user-attachments/assets/aa118358-f6fb-4576-b6a8-b62c189688f8)
![20241223_183459](https://github.com/user-attachments/assets/4c4ff607-cafe-40d6-b492-c40b5332ff11)
![20241223_183538](https://github.com/user-attachments/assets/79a16b47-3554-4998-9a44-b4662c5a09bb)
![20241223_183617](https://github.com/user-attachments/assets/41c48d1f-8d3e-4d5a-a52a-30d46680bdea)
![20241223_183727](https://github.com/user-attachments/assets/c8b63574-6ae4-48a5-aea7-c3533e32e466)

**Comment by si458** [comment]
so just an update, im having trouble trying to compile it
i managed to get it the toolchain for riscv eventually (i forgot to look at the nanokvm docs)
https://github.com/sipeed/NanoKVM/tree/main/server
but now its complaining about openssl
so i need to work out how to compile openssl for riscv
OR find the missing libcrypto.a and libssl.a for riscv64 from somewhere?

**Comment by leleb** [comment]
Worked on it and cross compiled openssl. Here the steps:

-In a Ubuntu Amd64 24.10 desktop virtual machine.
-install make

-download the riscv toolchain:
https://sophon-file.sophon.cn/sophon-prod-s3/drive/23/03/07/16/host-tools.tar.gz

( As referenced here:
https://github.com/sipeed/LicheeRV-Nano-Build/blob/v4.1.0/build/boards/cv181x/cv1812cp_licheerv_nano_sd/readme.md#compile-program-use-vendors-toolchain
  and here
https://github.com/sipeed/NanoKVM/tree/main/server
)

-extracted content (3 folders named riscv-*) and moved in a new folder named "TooolChains" in my home directory
added to path:
export PATH=$PATH:$HOME/ToolChains/riscv64-linux-musl-x86_64/bin

download OpenSSL_1_1_1-stable.zip from
https://github.com/openssl/openssl/tree/OpenSSL_1_1_1-stable
(https://openssl-library.org/source/)
  
extract and enter the extracted directory
cd OpenSSL_1_1_1-stable
./Configure linux-generic64 --cross-compile-prefix=riscv64-unknown-linux-musl- CFLAGS='-mcpu=c906fdv -march=rv64imafdcv0p7xthead -mcmodel=medany -mabi=lp64d'
  then
make

From now on you know how to:

-download meshagent sources and extract
copy the generated libcrypto.a and libssl.a from the OpenSSL_1_1_1-stable directory to MeshAgent/openssl/libstatic/linux/riscv64 

edit: MeshAgent/makefile

and add configuration 
PATH_RISCV64 = ../ToolChains/riscv64-linux-musl-x86_64/

...

ifeq ($(ARCHID),45)
ARCHNAME = riscv64
export PATH := $(PATH_RISCV64)bin:$(PATH_RISKV64)libexec/gcc/riscv64-unknown-linux-musl/10.2.0:$(PATH_RISCV64)riscv64-unknown-linux-musl/bin:$(PATH)
export STAGING_DIR := $(PATH_RISCV64)
CC = $(PATH_RISCV64)bin/riscv64-unknown-linux-musl-gcc
STRIP = $(PATH_RISCV64)bin/riscv64-unknown-linux-musl-strip
CEXTRA = -D_FORTIFY_SOURCE=2 -D_NOILIBSTACKDEBUG -D_NOFSWATCHER -Wformat -Wformat-security -fno-strict-aliasing -mcpu=c906fdv -march=rv64imafdcv0p7xthead -mcmodel=medany -mabi=lp64d
INCDIRS += -I$(PATH_RISCV64)include
KVM = 0
LMS = 0
endif


then
make linux ARCHID=45

**Comment by si458** [comment]
@leleb oh my god you had much more luck that i did!
i was bashing my head against the wall!
i will give you post a try tomorrow and then edit the meshcentral ui/code etc to include it if it works!

**Comment by si458** [comment]
@leleb i can indeed confirm your build instructions work!
and it does sorta work too on my nanokvm, it connects (takes about 3 mins dunno why)
then i can use the web http relay with mjpeg!
so its a full blown starting point!
would be amazing now if nanokvm added in a VNC server! then u could use the Web-VNC!

**Comment by leleb** [comment]
This looks promising https://libvnc.github.io/

**Comment by si458** [comment]
@leleb thanks, the is already issues open in the nanokvm to add in a vnc server, so nothing for us to do at the moment
https://github.com/sipeed/NanoKVM/issues/31
https://github.com/sipeed/NanoKVM/issues/181

**Comment by leleb** [comment]
I did not experienced a such long time (3min) for connecting. Except the first time the agent is started which takes a few seconds generating the keys, then next times is is immediately connecting. However I run it by command line starting the agent with the .msh in same folder.

**Comment by si458** [comment]
@leleb ahhh that might be what i was seeing then! as i was using `-connect` 
but indeed commands afterwards do indeed seem much quicker!
i also notice it didnt install the meshagent correctly but that might be the way the nanokvm is setup!
also i get `sh: getent: not found` 😆 but i think thats because the command `getent` doesnt exist on the nanokvm

**Comment by leleb** [comment]
@si458 yes I think install script must be revisited

**Comment by si458** [comment]
@leleb yes I've just spoke to Ylian and explained the binarys all work including the connect, but the install doesn't at the moment
So we need to have a rethink on how to do the install
Because it would require a recompile of ALL the linux meshagents and we don't have alot of the devices anymore to rebuild them
